pkg/executor/sortexec/sort.go

```diff
@@ -336,11 +480,115 @@ func (e *SortExec) fetchRowChunks(ctx context.Context) error {
 		return err
 	}
 
-	e.sortPartitions = append(e.sortPartitions, e.curPartition)
+	e.Unparallel.sortPartitions = append(e.Unparallel.sortPartitions, e.curPartition)
 	e.curPartition = nil
 	return nil
 }
 
+func (e *SortExec) fetchChunksParallel(ctx context.Context) error {
+	workerNum := len(e.Parallel.workers)
+
+	// Wait for the finish of all workers
+	workersWaiter := sync.WaitGroup{}
+	// Wait for the finish of chunk fetcher
+	fetcherWaiter := sync.WaitGroup{}
+
+	// Add before the start of goroutine to avoid that the counter is minus to negative.
+	workersWaiter.Add(workerNum)
+	fetcherWaiter.Add(1)
+
+	// Fetch chunks from child and put chunks into chunkChannel
+	go e.fetchChunksFromChild(ctx, &fetcherWaiter)
+
+	for i := range e.Parallel.workers {
+		e.Parallel.workers[i] = newParallelSortWorker(i, e.lessRow, e.Parallel.globalSortedRowsQueue, &workersWaiter, &e.Parallel.result, e.Parallel.chunkChannel, e.processErrorForParallel, e.finishCh, e.memTracker)
+		go e.Parallel.workers[i].run()
+	}
+
+	workersWaiter.Wait()
+
+	// Workers may panic, then no one could consume chunkChannel and
+	// the chunk fetcher may hang in this channel.
+	e.tryToCloseChunkChannel()
+
+	fetcherWaiter.Wait()
+
+	err := e.checkErrorForParallel()
+	if err != nil {
+		return err
+	}
+	e.fetchResultFromQueue()
+	return nil
+}
+
+// Fetch chunks from child and put chunks into chunkChannel
+func (e *SortExec) fetchChunksFromChild(ctx context.Context, waitGroup *sync.WaitGroup) {
+	defer func() {
+		if r := recover(); r != nil {
+			processPanicAndLog(e.processErrorForParallel, r)
+		}
+		e.tryToCloseChunkChannel()
+		waitGroup.Done()
+	}()
+
+	for {
+		chk := exec.TryNewCacheChunk(e.Children(0))
+		err := exec.Next(ctx, e.Children(0), chk)
+		if err != nil {
+			e.processErrorForParallel(err)
+			return
+		}
+
+		rowCount := chk.NumRows()
+		if rowCount == 0 {
+			break
+		}
+
+		e.memTracker.Consume(chk.MemoryUsage() + chunk.RowSize*int64(rowCount))
+
+		select {
+		case <-e.finishCh:
+			return
+		case e.Parallel.chunkChannel <- chk:
+			// chunkChannel may be closed in advance and it's ok to let it panic
+			// as workers have panicked and the query can't keep on running.
+		}
+
+		injectParallelSortRandomFail()
+	}
+}
+
+func (e *SortExec) checkErrorForParallel() error {
+	e.Parallel.errRWLock.RLock()
+	defer e.Parallel.errRWLock.RUnlock()
+	return e.Parallel.err
+}
+
+func (e *SortExec) processErrorForParallel(err error) {
+	e.Parallel.errRWLock.Lock()
+	defer e.Parallel.errRWLock.Unlock()
+	if e.Parallel.err == nil {
+		e.Parallel.err = err
+	}
+	e.tryToCloseFinishChannel()
```

Comment:
To reduce the scope of lock
```suggestion
        e.tryToCloseFinishChannel()
	e.Parallel.errRWLock.Lock()
	defer e.Parallel.errRWLock.Unlock()
	if e.Parallel.err == nil {
		e.Parallel.err = err
	}
```